### PR TITLE
Enable the inputs and outputs of the TE fusion group to support NHWC

### PR DESCRIPTION
**Pull request by EikanWang**
Fixes #ISSUE_NUMBER

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 1]
```diff
@@ -1172,11 +1198,12 @@ Tensor TensorExprKernel::convertSymbolicOutputToCorrectStrides(
           "Ouput tensor has no corresponding bufs in the fuser."));
   BufPtr buf = bufs_.at(v);
   // output is contiguous, no work to do
-  if (tensorOutputStrideDesc_[v->offset()] ==
-      torch::jit::StrideInput::TENSOR_CONT) {
+  auto stride_desc = tensorOutputStrideDesc_[v->offset()];
+  if (stride_desc == torch::jit::StrideInput::TENSOR_CONT ||
+      stride_desc == torch::jit::StrideInput::TENSOR_CONT_CHANNELS_LAST) {
```
Comment (jgong5) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1203]: If we can decide the contiguous property from the TE buffer directly, do we still need these two tensor-level descriptors to judge the contiguity?

File: torch/csrc/jit/tensorexpr/tensor.cpp [hunk 2]
```diff
@@ -39,12 +39,35 @@ StmtPtr Tensor::constructStmt(
...
+    if (buf_->is_contiguous()) {
+      for (int32_t i = args.size() - 1; i >= 0; i--) {
+        loop_order.push_back(i);
+      }
+    } else if (buf_->is_contiguous(c10::MemoryFormat::ChannelsLast)) {
+      loop_order = {1, 3, 2, 0};
+    } else if (buf_->is_contiguous(c10::MemoryFormat::ChannelsLast3d)) {
+      loop_order = {1, 4, 3, 2, 0};
+    } else {
+      loop_order = {1, 2, 0};
```
Comment (jgong5) [inline torch/csrc/jit/tensorexpr/tensor.cpp:59]: We assume it is always channels last 1D contiguous here? Why?

File: torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp [hunk 3]
```diff
@@ -164,8 +164,7 @@ StrideInput summarizeOutputStrides(const TensorType& tt) {
   // otherwise we defer to contiguous
   // TODO: channels last 3d
   // NNC Channels last permutation for outputs causes slowdown, disable
-  if (c10::is_channels_last_strides_2d(sizes, strides) &&
-      !tt.device()->is_cpu()) {
+  if (c10::is_channels_last_strides_2d(sizes, strides)) {
```
[the remark this answers is not in the document]
Comment (EikanWang) [reply torch/csrc/jit/passes/symbolic_shape_runtime_fusion.cpp:167]: It also works with 3d. But the StrideInput has not supported Channels-last 3D and 1D. I will enable 1D and 3D after the NHWC mechanism is work properly.
https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/symbolic_shape_runtime_fusion.h#L42

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 4]
```diff
@@ -999,19 +1001,39 @@ Tensor TensorExprKernel::bindInput(const torch::jit::Value* input) {
   auto const& outputs = input->owningGraph()->outputs();
   std::unordered_set<const Value*> outputs_set(outputs.begin(), outputs.end());
 
+  auto is_static_cont = [](const torch::jit::Value* input) {
+    if (input->isCompleteTensor()) {
+      return isContiguous(input) ||
+          isContiguous(input, at::MemoryFormat::ChannelsLast);
```
[the remark this answers is not in the document]
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1007]: Same as above.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 1 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1203]: We use the `StrideInput` for the inputs and outputs of the NNC fusion group graph, while the `ExprHandle` is used to describe the `Buf` within the fusion graph. The previous is for the graph boundary, the inputs and outputs are `jit::Value`. The next one is for the graph internal, the data is `Buf`

File: torch/csrc/jit/tensorexpr/tensor.cpp (hunk 2 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/tensor.cpp:59]: Currently, we only support contiguous tensors. The `bindInput` converts the non-contiguous tensors to contiguous.

Comment (jgong5) [reply torch/csrc/jit/tensorexpr/tensor.cpp:59]: Should we put some assertion here just in case the assumption changes in the future?

Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/tensor.cpp:59]: Line 42-46 is the assertion.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 1 above)
Comment (jgong5) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1203]: Does NNC supports non-contiguous tensors at the TorchScript JIT graph boundary? If so, how TE handles such inputs?

File: torch/csrc/jit/tensorexpr/tensor.cpp (hunk 2 above)
Comment (jgong5) [reply torch/csrc/jit/tensorexpr/tensor.cpp:59]: Oh, I didn't notice it. Thanks.

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 1 above)
Comment (EikanWang) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1203]: Yes, NNC does. A dedicated conversion statement will be added to convert the tensor to contiguous. https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/kernel.cpp#L1045

Comment (jgong5) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1203]: Make sense to assume some form of contiguity at the boundary. Thanks.

**Review by jgong5 — APPROVED** [review]

**Pull request closed without merging**